### PR TITLE
[extractor/generic]  add full support for HTML5 media tags(extract single and multiple formats, subtitles, poster...)

### PR DESCRIPTION
**Pull request by remitamine**
before:
```
youtube-dl -F http://www.dota2.com/reborn/part1/
[generic] part1: Requesting header
WARNING: Falling back on generic information extractor.
[generic] part1: Downloading webpage
[generic] part1: Extracting information
[download] Downloading playlist: None
[generic] playlist None: Collected 4 video ids (downloading 4 of them)
[download] Downloading video 1 of 4
[info] Available formats for AllNewDashboard:
format code  extension  resolution note
0            webm       unknown    
[download] Downloading video 2 of 4
[info] Available formats for RedesignedHeroBrowser:
format code  extension  resolution note
0            webm       unknown    
[download] Downloading video 3 of 4
[info] Available formats for NewWaysToWatch:
format code  extension  resolution note
0            webm       unknown    
[download] Downloading video 4 of 4
[info] Available formats for RevampedLearningExperience:
format code  extension  resolution note
0            webm       unknown
```
after:
```
youtube-dl -F http://www.dota2.com/reborn/part1/
[generic] part1: Requesting header
WARNING: Falling back on generic information extractor.
[generic] part1: Downloading webpage
[generic] part1: Extracting information
[download] Downloading playlist: None
[generic] playlist None: Collected 4 video ids (downloading 4 of them)
[download] Downloading video 1 of 4
[info] Available formats for :
format code  extension  resolution note
0            webm       unknown    
1            mp4        unknown    (best)
[download] Downloading video 2 of 4
[info] Available formats for :
format code  extension  resolution note
0            webm       unknown    
1            mp4        unknown    (best)
[download] Downloading video 3 of 4
[info] Available formats for :
format code  extension  resolution note
0            webm       unknown    
1            mp4        unknown    (best)
[download] Downloading video 4 of 4
[info] Available formats for :
format code  extension  resolution note
0            webm       unknown    
1            mp4        unknown    (best)
```

### CODE REVIEW

File: youtube_dl/extractor/generic.py [hunk 1]
```diff
@@ -1820,33 +1822,41 @@ def filter_video(urls):
...
-        for video_url in found:
-            video_url = compat_urlparse.urljoin(url, video_url)
-            video_id = compat_urllib_parse_unquote(os.path.basename(video_url))
-
-            # Sometimes, jwplayer extraction will result in a YouTube URL
-            if YoutubeIE.suitable(video_url):
-                entries.append(self.url_result(video_url, 'Youtube'))
-                continue
+        for video_urls in found:
+            if isinstance(video_urls, str):
```
Comment (jaimeMF) [inline youtube_dl/extractor/generic.py]: `compat_str`, on python 2.x `str` is the same as `bytes`.

**Comment by remitamine** [comment]
> compat_str, on python 2.x str is the same as bytes.
i changed the code now to use compat_str.

**Comment by dstftw** [comment]
Add test.

**Comment by remitamine** [comment]
this are some changes i want to add, i will put the diff here to discuss before i commit them.
```
diff --git a/youtube_dl/extractor/generic.py b/youtube_dl/extractor/generic.py
index 08c5fdb..e2487c3 100644
--- a/youtube_dl/extractor/generic.py
+++ b/youtube_dl/extractor/generic.py
@@ -1013,7 +1013,40 @@ class GenericIE(InfoExtractor):
                 'ext': 'mp4',
                 'title': 'cinemasnob',
             },
-        }
+        },
+        # HTML5 Videos with multiple formats
+        {
+            'url': 'https://commons.wikimedia.org/wiki/Big_Buck_Bunny',
+            'info_dict': {
+                'id': 'Big Buck Bunny - Wikimedia Commons',
+                'title': 'Big Buck Bunny - Wikimedia Commons',
+            },
+            'playlist': [{
+                'md5': '78467f74f821d12f22843647a9017e1a',
+                'info_dict': {
+                    'id': 'Big_Buck_Bunny_small',
+                    'ext': 'webm',
+                    'title': 'Big_Buck_Bunny_small (1)',
diff --git a/youtube_dl/extractor/generic.py b/youtube_dl/extractor/generic.py
index 08c5fdb..e2487c3 100644
--- a/youtube_dl/extractor/generic.py
+++ b/youtube_dl/extractor/generic.py
@@ -1013,7 +1013,40 @@ class GenericIE(InfoExtractor):
                 'ext': 'mp4',
                 'title': 'cinemasnob',
             },
-        }
+        },
+        # HTML5 Videos with multiple formats
+        {
+            'url': 'https://commons.wikimedia.org/wiki/Big_Buck_Bunny',
+            'info_dict': {
+                'id': 'Big Buck Bunny - Wikimedia Commons',
+                'title': 'Big Buck Bunny - Wikimedia Commons',
+            },
+            'playlist': [{
+                'md5': '78467f74f821d12f22843647a9017e1a',
+                'info_dict': {
+                    'id': 'Big_Buck_Bunny_small',
+                    'ext': 'webm',
+                    'title': 'Big_Buck_Bunny_small (1)',
+                    'uploader': 'commons.wikimedia.org',
+                },
+            }, {
+                'md5': 'efab0fd5dfe10767df1ff5d923adc1d5',
+                'info_dict': {
+                    'id': 'Big_Buck_Bunny_medium.ogv.480p',
+                    'ext': 'webm',
+                    'title': 'Big_Buck_Bunny_medium.ogv.480p (2)',
+                    'uploader': 'commons.wikimedia.org',
+                },
+            }, {
+                'md5': '57495cddd8213e107e9227ed738bd26b',
+                'info_dict': {
+                    'id': 'Big_Buck_Bunny_8_seconds_bird_clip.ogv.720p',
+                    'ext': 'webm',
+                    'title': 'Big_Buck_Bunny_8_seconds_bird_clip.ogv.720p (3)',
+                    'uploader': 'commons.wikimedia.org',
+                },
+            }],
+        },
     ]
 
     def report_following_redirect(self, new_url):
@@ -1827,7 +1860,7 @@ class GenericIE(InfoExtractor):
             if isinstance(video_urls, compat_str):
                 video_urls = [video_urls]
 
-            video_id = compat_urllib_parse_unquote(os.path.basename(url))
+            video_id = compat_urllib_parse_unquote(os.path.basename(video_urls[0]))
             # here's a fun little line of code for you:
             video_id = os.path.splitext(video_id)[0]
+                },
+            }, {
+                'md5': '57495cddd8213e107e9227ed738bd26b',
+                'info_dict': {
+                    'id': 'Big_Buck_Bunny_8_seconds_bird_clip.ogv.720p',
+                    'ext': 'webm',
+                    'title': 'Big_Buck_Bunny_8_seconds_bird_clip.ogv.720p (3)',
+                    'uploader': 'commons.wikimedia.org',
+                },
+            }],
+        },
     ]
 
     def report_following_redirect(self, new_url):
@@ -1827,7 +1860,7 @@ class GenericIE(InfoExtractor):
             if isinstance(video_urls, compat_str):
                 video_urls = [video_urls]
 
-            video_id = compat_urllib_parse_unquote(os.path.basename(url))
+            video_id = compat_urllib_parse_unquote(os.path.basename(video_urls[
0]))
             # here's a fun little line of code for you:
             video_id = os.path.splitext(video_id)[0]
 
@@ -1855,11 +1888,13 @@ class GenericIE(InfoExtractor):
                     formats.append({'url': video_url})
 
             if formats:
+                if len(formats) > 0:
+                    self._sort_formats(formats)
                 entries.append({
                     'id': video_id,
                     'formats': formats,
                     'uploader': video_uploader,
-                    'title': video_title,
+                    'title': video_id,
                     'age_limit': age_limit,
                 })
 
@@ -1873,4 +1908,6 @@ class GenericIE(InfoExtractor):
             return {
                 '_type': 'playlist',
                 'entries': entries,
+                'id': video_title,
+                'title': video_title,
             }
```

**Comment by dstftw** [comment]
Just commit it. There is no way to add a note to lines of codeblock in comment therefore no proper way to discuss code. Moreover this involves applying patch instead of just pulling the branch.

**Comment by remitamine** [comment]
ok i will comment when i'm done, i want to add full support for video tags there are cases are not handled by this code like these case:
```html
<video src="http://v2v.cc/~j/theora_testsuite/320x240.ogg" controls>
  Your browser does not support the <code>video</code> element.
</video>
```
and extracting metadata if it is available.

**Comment by remitamine** [comment]
i added support for inline src videos,subtitles and poster and also a test.

File: youtube_dl/extractor/generic.py [hunk 2]
```diff
@@ -1796,8 +1830,50 @@ def filter_video(urls):
...
-            # HTML5 video
-            found = re.findall(r'(?s)<(?:video|audio)[^<]*(?:>.*?<source[^>]*)?\s+src=["\'](.*?)["\']', webpage)
+            # HTML5 media(video or audio)
+            found = re.findall(r'(?s)<(video|audio)([^>]*)>(.*?)</(?:video|audio)>', webpage)
+            if found:
+                def extract_attributes(attributes_str, attributes_regex=r'(?s)\s*([^\s=]+)\s*=\s*["\']([^"\']+)["\']'):
+                    attributes = re.findall(attributes_regex, attributes_str)
+                    attributes_dict = {}
+                    if attributes:
+                        attributes_dict = {attribute_name: attribute_value for (attribute_name, attribute_value) in attributes}
```
Comment (dstftw) [inline youtube_dl/extractor/generic.py]: dict comprehensions don't work in python 2.6.

**Comment by dstftw** [comment]
I guess it's reasonable to extract html5 media processing into separate method for clarity and reusability similarly to SMIL, XSPF and so.

**Comment by remitamine** [comment]
> dict comprehensions don't work in python 2.6.

i will fix it.
> I guess it's reasonable to extract html5 media processing into separate method for clarity and reusability similarly to SMIL, XSPF and so.

i will work on it.
i think also that extract_attributes can be reused, maybe utils.py is a good place for it.

**Comment by remitamine** [comment]
i made the changes there one thing i want ask i use the extract_attributes in two pull requests(this and and #6828) i think this will make a conflict or duplication  where i should keep it here or in the other pull request.

**Comment by remitamine** [comment]
will start the work in a new PR using the new extract_attributes function.

**Pull request closed without merging**